metricproviders/kayenta/kayenta.go

```diff
@@ -0,0 +1,202 @@
+	"errors"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"io/ioutil"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+
+	metricutil "github.com/argoproj/argo-rollouts/utils/metric"
+	templateutil "github.com/argoproj/argo-rollouts/utils/template"
+)
+
+const (
+	//ProviderType indicates the provider is kayenta
+	ProviderType = "Kayenta"
+	KayentaScoreURL string = "{{inputs.address}}/canary/{{inputs.canaryExecutionId}}"
+
+	JobURL string = "{{inputs.address}}/canary/{{inputs.canaryConfigId}}?application={{inputs.application}}&metricsAccountName={{inputs.metricsAccountName}}&configurationAccountName={{inputs.configurationAccountName}}&storageAccountName={{inputs.storageAccountName}}"
+
+	JobPayloadTemplate string = `
+							{
+								"scopes": {
+										{{inputs.scopes}}
+								},
+                                "thresholds" : {
+                                    "pass": {{inputs.pass}},
+                                    "marginal": {{inputs.marginal}}
+                                }
+                            }`
+
+
+)
+
+type Provider struct {
+	logCtx log.Entry
+	client http.Client
+}
+
+// Type incidates provider is a kayenta provider
+func (p *Provider) Type() string {
+	return ProviderType
+}
+
+// Run queries kayentd for the metric
+func (p *Provider) Run(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric) v1alpha1.Measurement {
+	startTime := metav1.Now()
+	newMeasurement := v1alpha1.Measurement{
+		StartedAt: &startTime,
+	}
+
+	jobURL := strings.Replace(JobURL, "{{inputs.address}}", metric.Provider.Kayenta.Address, 1)
+	jobURL = strings.Replace(jobURL, "{{inputs.canaryConfigId}}", metric.Provider.Kayenta.CanaryConfigId, 1)
+	jobURL = strings.Replace(jobURL, "{{inputs.application}}", metric.Provider.Kayenta.Application, 1)
+	jobURL = strings.Replace(jobURL, "{{inputs.metricsAccountName}}", metric.Provider.Kayenta.MetricsAccountName, 1)
+	jobURL = strings.Replace(jobURL, "{{inputs.configurationAccountName}}", metric.Provider.Kayenta.ConfigurationAccountName, 1)
+	jobURL = strings.Replace(jobURL, "{{inputs.storageAccountName}}", metric.Provider.Kayenta.StorageAccountName, 1)
+
+	jobPayLoad := strings.Replace(JobPayloadTemplate, "{{inputs.pass}}", strconv.Itoa(metric.Provider.Kayenta.Threshold.Pass), 1)
+	jobPayLoad = strings.Replace(jobPayLoad, "{{inputs.marginal}}", strconv.Itoa(metric.Provider.Kayenta.Threshold.Marginal), 1)
+	var scopes string
+	for i, s := range metric.Provider.Kayenta.Scopes {
+		name := s.Name
+		controlScope, err := json.Marshal(s.ControlScope)
+		if err != nil {
+			return metricutil.MarkMeasurementError(newMeasurement, err)
+		}
+		experimentScope, err := json.Marshal(s.ExperimentScope)
+		if err != nil {
+			return metricutil.MarkMeasurementError(newMeasurement, err)
+		}
+		scopes = "\"" + name +  "\":" + string(controlScope) + "," + string(experimentScope)
```

Comment:
Is this a valid key pair? I think the key is `"\"" + name +  "\":"` but don't we need brackets around `string(controlScope) + "," + string(experimentScope)` to make it the value?

I imagined that an indiviudal scope would be constructed something like this:
```
const scopeFormat = `"%s":{"controlScope": %s, "experimentScope": %s}`
scope := fmt.Sprintf(scopeFormat, name, controlScope, experimentScope)
```
Is there something that I am missing?

Also, will we lose the first scope if we have a second scope?

Reply:
fixed.  I had scopes = scopes + <next scope>